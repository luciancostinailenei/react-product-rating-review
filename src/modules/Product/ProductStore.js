import { action, computed, observable } from 'mobx'

import moment from 'moment'

class ProductStore {
  @observable reviews = []
  @observable isProductReviewFormDisplayed = false

  @action
  addReview = review => {
    this.reviews.push(review)
  }

  @computed
  get reviewsCount() {
    return this.reviews.length
  }

  @computed
  get sortedReviews() {
    const sortedReviews = this.reviews.sort((a, b) => {
      if (moment(a.date) > moment(b.date)) {
        return -1
      }

      if (moment(a.date) < moment(b.date)) {
        return 1
      }

      return 0
    })

    return sortedReviews
  }

  @computed
  get averageScore() {
    const numberOfReviews = this.reviewsCount

    if (numberOfReviews === 0) {
      return 0
    }

    const scoreSum = this.reviews.reduce((acc, review) => {
      return acc + review.rating
    }, 0)

    const averageScore = scoreSum / numberOfReviews
    return averageScore % 1 === 0 ? averageScore : averageScore.toFixed(1)
  }
}

const store = new ProductStore()
export default store
