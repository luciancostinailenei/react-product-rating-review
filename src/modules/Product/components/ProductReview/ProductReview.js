import React from 'react'
import Moment from 'moment'

import './ProductReview.scss'

import ProductRating from '../ProductRating/ProductRating'

const ProductReview = props => {
  return (
    <li className="ProductReview">
      <h4 className="ProductReview__title">{props.reviewData.title}</h4>
      <div className="ProductReview__rating">
        <ProductRating
          inputIdIndex={props.index}
          scoreValue={props.reviewData.rating}
        />
        <div className="ProductReview__info">
          <span className="ProductReview__user-name">
            {props.reviewData.user}
          </span>
          &nbsp; | &nbsp;
          <span className="ProductReview__date">
            {Moment(props.reviewData.date).format('DD/MM/YYYY')}
          </span>
        </div>
      </div>
      <div />
      <p className="ProductReview__text">{props.reviewData.text}</p>
    </li>
  )
}

export default ProductReview
