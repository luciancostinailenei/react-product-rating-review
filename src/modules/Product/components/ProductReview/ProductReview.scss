/*------------------------------------*\
  #ProductReview
\*------------------------------------*/

.ProductReview { 
    border-bottom: 1px solid var(--gray-300);
    padding-bottom: 10px; 
    padding-left: 10px;

    &:last-of-type {
        border-bottom: none;
    }
}

    .ProductReview__title { 
        font-size: 15px; 
        margin-bottom: 0px;
    }

    .ProductReview__rating { 
        align-items: center;
        display: flex;
        margin-top: 5px; 
    }

    .ProductReview__info {
        margin-left: 10px;
    }

    .ProductReview__user-name {
        color: var(--teal-400);
    }