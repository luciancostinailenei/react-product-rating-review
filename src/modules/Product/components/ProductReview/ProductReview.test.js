import { configure, shallow } from 'enzyme'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import toJson, { createSerializer } from 'enzyme-to-json'

import ProductReview from './ProductReview'

import ProductStore from '../../ProductStore'

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

configure({ adapter: new Adapter() })

describe('ProductReview', () => {
  it('matches snapshot', () => {
    const mockReviewData = {
      rating: 5,
      title: 'Test Review Title 1',
      text: 'Review Text Test 1',
      user: 'Lucian',
      date: 'Wed Mar 14 2018 18:58:29 GMT+0200 (GTB Standard Time)'
    }

    const ProductReviewsWrapper = shallow(
      <ProductReview
        ProductStore={ProductStore}
        reviewData={mockReviewData}
        key={'1'}
        index={'1'}
      />
    )

    expect(toJson(ProductReviewsWrapper)).toMatchSnapshot()
  })

  it('renders accordingly based on given props', () => {
    const mockReviewData = {
      rating: 5,
      title: 'Test Review Title 1',
      text: 'Review Text Test 1',
      user: 'Lucian',
      date: 'Wed Mar 14 2018 18:58:29 GMT+0200 (GTB Standard Time)'
    }

    const ProductReviewWrapper = shallow(
      <ProductReview
        ProductStore={ProductStore}
        reviewData={mockReviewData}
        key={'1'}
        index={'1'}
      />
    )

    expect(ProductReviewWrapper.find('.ProductReview__title').text()).toBe(
      'Test Review Title 1'
    )

    expect(ProductReviewWrapper.find('.ProductReview__user-name').text()).toBe(
      'Lucian'
    )

    expect(ProductReviewWrapper.find('.ProductReview__date').text()).toBe(
      '14/03/2018'
    )

    expect(ProductReviewWrapper.find('.ProductReview__text').text()).toBe(
      'Review Text Test 1'
    )
  })
})
