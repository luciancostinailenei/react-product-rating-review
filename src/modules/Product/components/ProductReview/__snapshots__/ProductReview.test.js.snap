// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductReview matches snapshot 1`] = `
<li
  className="ProductReview"
>
  <h4
    className="ProductReview__title"
  >
    Test Review Title 1
  </h4>
  <div
    className="ProductReview__rating"
  >
    <inject-ProductRating-with-ProductStore
      inputIdIndex="1"
      scoreValue={5}
    />
    <div
      className="ProductReview__info"
    >
      <span
        className="ProductReview__user-name"
      >
        Lucian
      </span>
        |  
      <span
        className="ProductReview__date"
      >
        14/03/2018
      </span>
    </div>
  </div>
  <div />
  <p
    className="ProductReview__text"
  >
    Review Text Test 1
  </p>
</li>
`;
